updater/lib/dependabot/update_graph_processor.rb

```diff
@@ -0,0 +1,125 @@
+# and submitting a snapshot of each to the GitHub Dependency Submission API.
+module Dependabot
+  class UpdateGraphProcessor
+    extend T::Sig
+
+    # To do work, this class needs three arguments:
+    # - The Dependabot::Service to send events and outcomes to
+    # - The Dependabot::Job that describes the work to be done
+    # - The Dependabot::DependencyFile list retrieved by the file fetcher
+    # - The base_commit_sha being processed
+    sig do
+      params(
+        service: Dependabot::Service,
+        job: Dependabot::Job,
+        base_commit_sha: String,
+        dependency_files: T::Array[Dependabot::DependencyFile]
+      ).void
+    end
+    def initialize(service:, job:, base_commit_sha:, dependency_files:)
+      @service = service
+      @job = job
+      @base_commit_sha = base_commit_sha
+      @dependency_files = dependency_files
+
+      @error_handler = T.let(
+        Dependabot::Updater::ErrorHandler.new(service: service, job: job),
+        Dependabot::Updater::ErrorHandler
+      )
+    end
+
+    sig { void }
+    def run
+      raise Dependabot::DependabotError, "job.source.directories is nil" if job.source.directories.nil?
+      raise Dependabot::DependabotError, "job.source.directories is empty" unless job.source.directories&.any?
+      raise Dependabot::DependabotError, "job.source.branch is nil" if job.source.branch.nil?
+
+      T.must(job.source.directories).each do |directory|
+        directory_source = create_source_for(directory)
+        directory_dependency_files = dependency_files_for(directory)
+
+        submission = create_submission(directory_source, directory_dependency_files)
+
+        Dependabot.logger.info("Dependency submission payload:\n#{JSON.pretty_generate(submission.payload)}")
+        service.create_dependency_submission(dependency_submission: submission)
+      end
+    end
+
+    private
+
+    sig { returns(Dependabot::Service) }
+    attr_reader :service
+
+    sig { returns(Dependabot::Job) }
+    attr_reader :job
+
+    sig { returns(String) }
+    attr_reader :base_commit_sha
+
+    sig { returns(T::Array[Dependabot::DependencyFile]) }
+    attr_reader :dependency_files
+
+    sig { returns(Dependabot::Updater::ErrorHandler) }
+    attr_reader :error_handler
+
+    sig { params(directory: String).returns(Dependabot::Source) }
+    def create_source_for(directory)
+      job.source.dup.tap do |s|
+        s.directory = directory
+      end
+    end
+
+    sig { params(directory: String).returns(T::Array[Dependabot::DependencyFile]) }
+    def dependency_files_for(directory)
+      dependency_files.select { |f| f.directory == directory }
+    end
+
+    sig do
+      params(
+        source: Dependabot::Source,
+        files: T::Array[Dependabot::DependencyFile]
+      ).returns(GithubApi::DependencySubmission)
+    end
+    def create_submission(source, files)
+      # TODO(brrygrdn): Refactor the grapher to wrap the parser call
```

Comment:
This is the difficult change a lot of this PR is in service of making easy to pull off, I plan to address it in a follow up as it'll require an interface change running through `common` and the Updater so I'dd rather do it separately.